internal/core/drand_beacon_public.go

```diff
@@ -55,43 +57,55 @@ func (bp *BeaconProcess) PublicRand(ctx context.Context, in *drand.PublicRandReq
 	beaconResp, err := bp.beacon.Store().Last(ctx)
 	if wanted := in.GetRound(); err == nil && wanted == beaconResp.GetRound()+1 {
 		// we got a request for the next round about to be produced, let's honor it with a callback
-		// we cancel after period since we should never wait so long
-		ctx, cancel := context.WithTimeout(ctx, bp.group.Period)
-		defer cancel()
+		// to make sure we don't miss it because of the aggregation time
+		cctx, cancel := context.WithCancel(ctx)
 		waitlist := make(chan *common.Beacon, 1)
 		rnd := make([]byte, sha256.Size)
 		_, err = rand.Read(rnd)
 		if err != nil {
-			rnd = beaconResp.GetRandomness()[:sha256.Size]
+			_ = copy(rnd, beaconResp.GetRandomness())
 		}
-		fn := func(b *common.Beacon, closed bool) {
-			if closed {
-				close(waitlist)
+		var mu sync.Mutex
```

Comment:
this mutex gets locked in `fn`, but calling `PublicRand` again will just create a new mutex so it doesn't really protect anything

Reply:
okay right I understand why this mutex works now